File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.java
class Solution {
    List<Integer> path = new LinkedList<>();
    List<List<Integer>> ans = new ArrayList<>();
    
    public List<List<Integer>> findSubsequences(int[] nums) {
        backtracking(nums, 0);
        return ans;
    }
    
    private void backtracking(int[] nums, int startIdx) {
        if (path.size() >= 2) {
            ans.add(new ArrayList<>(path));
        }
        
        Set<Integer> used = new HashSet<>();
        
        for (int i = startIdx; i < nums.length; i++) {
            if (path.size() > 0 && nums[i] < path.get(path.size() - 1) || used.contains(nums[i])) {
                continue;
            }
            
            path.add(nums[i]);
            used.add(nums[i]);
            backtracking(nums, i + 1);
            path.removeLast();
        }
    }
}

File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.py
class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def backtracking(path: List[int], index: int) -> None:
            if len(path) >= 2:
                ans.append(path[:])
            
            used = set()

            for i in range(index, len(nums)):
                if path and nums[i] < path[-1] or nums[i] in used:
                    continue
                
                path.append(nums[i])
                used.add(nums[i])
                backtracking(path, i + 1)
                path.pop()
        
        backtracking([], 0)
        return ans